libsolidity/codegen/CompilerUtils.cpp

```diff
@@ -1202,9 +1206,38 @@ void CompilerUtils::computeHashStatic()
 
 void CompilerUtils::storeStringData(bytesConstRef _data)
 {
-	//@todo provide both alternatives to the optimiser
+	// @todo provide both alternatives to the optimiser
 	// stack: mempos
-	if (_data.size() <= 128)
+	bool storeAsData = false;
+	std::cout << "storeStringData: dataSize=" << _data.size() << std::endl;
+	if (_data.size() > 32)
+	{
+		// @todo this does not take the constant optimiser into account,
+		// which could reduce the costs of individual PUSHes
+		eth::AssemblyItems pushItems;
+		unsigned numWords = (_data.size() + 31) / 32;
+		for (unsigned i = 0; i < numWords; i++)
+		{
+			pushItems.push_back(eth::AssemblyItem(u256(-1))); // value (assume all bits set)
+			pushItems.push_back(eth::AssemblyItem(u256(4096 + i * 32))); // memory offset
+			pushItems.push_back(eth::AssemblyItem(Instruction::MSTORE));
+		}
+		auto pushCost = eth::PathGasMeter::estimateMax(pushItems, m_context.evmVersion(), 0, make_shared<eth::KnownState>());
+
+		auto copyCost = eth::PathGasMeter::estimateMax(eth::AssemblyItems{
+			u256(_data.size()), // length
+			u256(4096), // memory offset, assume memory is this big
+			u256(4096), // code offset, assume code is this big
+			Instruction::CODECOPY
+		}, m_context.evmVersion(), 0, make_shared<eth::KnownState>());
+
+		std::cout << "storeStringData: dataSize=" << _data.size() << " pushCost=" << pushCost << " copyCost=" << copyCost << std::endl;
+
+		if (copyCost < pushCost)
```

Comment:
This should also use the optimizer tradeoff parameter.

Reply:
I'm not sure I fully agree.  Using `PUSH` after the threshold will always be more expensive, because: it is more bytecode.